packages/utils/src/time.ts

```diff
@@ -1,98 +1,154 @@
+  //
+  // This is a partial workaround for browsers reporting performance.timeOrigin such that performance.timeOrigin +
+  // performance.now() gives a date arbitrarily in the past.
+  //
+  // Additionally, computing timeOrigin in this way fills the gap for browsers where performance.timeOrigin is
+  // undefined.
+  //
+  // The assumption that performance.timeOrigin + performance.now() ~= Date.now() is flawed, but we depend on it to
+  // interact with data coming out of performance entries.
+  //
+  // Note that despite recommendations against it in the spec, browsers implement the Performance API with a clock that
+  // might stop when the computer is asleep (and perhaps under other circumstances). Such behavior causes
+  // performance.timeOrigin + performance.now() to have an arbitrary skew over Date.now(). In laptop computers, we have
+  // observed skews that can be as long as days, weeks or months.
+  //
+  // See https://github.com/getsentry/sentry-javascript/issues/2590.
+  //
+  // BUG: despite our best intentions, this workaround has its limitations. It mostly addresses timings of pageload
+  // transactions, but ignores the skew built up over time that can aversely affect timestamps of navigation
+  // transactions of long-lived web pages.
+  const timeOrigin = Date.now() - performance.now();
 
-const performanceFallback: CrossPlatformPerformance = {
-  now(): number {
-    let now = Date.now() - INITIAL_TIME;
-    if (now < prevNow) {
-      now = prevNow;
-    }
-    prevNow = now;
-    return now;
-  },
-  timeOrigin: INITIAL_TIME,
-};
+  return {
+    now: () => performance.now(),
+    timeOrigin,
+  };
+}
 
-const crossPlatformPerformance: CrossPlatformPerformance = ((): CrossPlatformPerformance => {
-  // React Native's performance.now() starts with a gigantic offset, so we need to wrap it.
-  if (isReactNative()) {
-    return getReactNativePerformanceWrapper();
+/**
+ * Returns the native Performance API implementation from Node.js. Returns undefined in old Node.js versions that don't
+ * implement the API.
+ */
+function getNodePerformance(): Performance | undefined {
+  try {
+    const perfHooks = dynamicRequire(module, 'perf_hooks') as { performance: Performance };
+    return perfHooks.performance;
+  } catch (_) {
+    return undefined;
   }
+}
 
+/**
+ * The Performance API implementation for the current platform, if available.
+ */
+const platformPerformance = ((): Performance | undefined => {
   if (isNodeEnv()) {
-    try {
-      const perfHooks = dynamicRequire(module, 'perf_hooks') as { performance: CrossPlatformPerformance };
-      return perfHooks.performance;
-    } catch (_) {
-      return performanceFallback;
-    }
+    return getNodePerformance();
   }
+  return getBrowserPerformance();
+})();
 
-  const { performance } = getGlobalObject<Window>();
-
-  if (!performance || !performance.now) {
-    return performanceFallback;
-  }
+let timestampSource = dateTimestampSource;
```

Comment:
I'm pretty sure we can just move `dateTimestampSource` down here and assign it conditionally.